ChemistryLib/PhreeqcKernel.cpp

```diff
@@ -0,0 +1,257 @@
+
+    // solution
+    for (std::size_t chemical_system_id = 0;
+         chemical_system_id < num_chemical_systems;
+         ++chemical_system_id)
+    {
+        aqueous_solution.setChemicalSystemID(chemical_system_id);
+        Rxn_solution_map.emplace(chemical_system_id,
+                                 *aqueous_solution.castToBaseClass());
+    }
+    use.Set_solution_in(true);
+
+    // kinetics
+    if (kinetic_reactants->isKineticReactantDefined())
+    {
+        for (std::size_t chemical_system_id = 0;
+             chemical_system_id < num_chemical_systems;
+             ++chemical_system_id)
+        {
+            kinetic_reactants->setChemicalSystemID(chemical_system_id);
+            Rxn_kinetics_map.emplace(chemical_system_id,
+                                     *kinetic_reactants->castToBaseClass());
+        }
+        use.Set_kinetics_in(true);
+    }
+
+    reinitializeRates();
+
+    setConvergenceTolerance();
+
+    configureOutputSettings();
+
+    for (auto const& map_pair : process_id_to_component_name_map)
+    {
+        auto const& transport_process_id = map_pair.first;
+        auto const& transport_process_variable = map_pair.second;
+
+        auto master_species =
+            master_bsearch(transport_process_variable.c_str());
+
+        _process_id_to_master_map[transport_process_id] = master_species;
+    }
+}
+
+void PhreeqcKernel::reinitializeRates()
+{
+    count_rates = _reaction_rates.size();
+    rates = (struct rate*)realloc(
+        rates, (std::size_t)(count_rates) * sizeof(struct rate));
+    int rate_id = 0;
+    for (auto const& reaction_rate : _reaction_rates)
+    {
+        rates[rate_id].name = reaction_rate.kinetic_reactant.data();
+
+        // Commands' strings are freed by Phreeqc dtor.
+        rates[rate_id].commands = static_cast<char*>(
+            malloc(sizeof(char) * reaction_rate.commands().size() + 1));
+        if (rates[rate_id].commands == nullptr)
+            OGS_FATAL("Could not allocate memory for rate[%d] commands.",
+                      rate_id);
+        reaction_rate.commands().copy(rates[rate_id].commands,
+                                      std::string::npos);
+        rates[rate_id].commands[reaction_rate.commands().size()] = '\0';
+
+        rates[rate_id].new_def = 1;
+        rates[rate_id].linebase = nullptr;
+        rates[rate_id].varbase = nullptr;
+        rates[rate_id].loopbase = nullptr;
+
+        ++rate_id;
+    };
+}
+
+void PhreeqcKernel::doWaterChemistryCalculation(
+    std::vector<GlobalVector*>& process_solutions, double const dt)
+{
+    setAqueousSolutions(process_solutions);
+    setTimeStepSize(dt);
+
+    execute(process_solutions);
+}
+
+void PhreeqcKernel::setAqueousSolutions(
+    std::vector<GlobalVector*> const& process_solutions)
+{
+    // Loop over chemical systems
+    auto const num_chemical_systems = process_solutions[0]->size();
+    for (auto chemical_system_id = 0; chemical_system_id < num_chemical_systems;
```

Comment:
This will complain about mismatched integer sizes: `num_chemical_systems` is most likely a `std::size_t`, and `chemical_system_id` is an `int` because the `0` is.

Reply:
Fixed here and elsewhere.